File: runtime/cpp/titan-ast-runtime-gui/gui/AstGuiOutputer.cpp
//
// Created by tian wei jun on 2022/11/24 0024.
//

#include "AstGuiOutputer.h"
#include "GuiApi.h"

AstGuiOutputer::AstGuiOutputer(const Ast *ast) : ast(ast), stringTree(nullptr), hasOpened(false) {
}
AstGuiOutputer::~AstGuiOutputer() {
  delete stringTree;
  stringTree = nullptr;
}

void AstGuiOutputer::output() {
  if (!ast) {
    return;
  }
  StringTree *ast2StringTree = buildStringTree(ast);
  futureOfView =
      std::async(std::launch::async, guiapi::showViewUntilClose, ast2StringTree);
  hasOpened = true;
}
void AstGuiOutputer::waitToClose() {
  if (hasOpened) {
    futureOfView.get();
    hasOpened = false;
  }
}

StringTree *AstGuiOutputer::buildStringTree(const Ast *argAst) {
  if (!argAst) {
    return nullptr;
  }
  auto *strTree = new StringTree();
  strTree->text = argAst->toString();
  if (!argAst->children.empty()) {
    for (Ast *astTreeChild : argAst->children) {
      StringTree *strTreeChild = buildStringTree(astTreeChild);
      strTree->children->push_back(strTreeChild);
    }
  }
  return strTree;
}

File: runtime/java/titan-ast-runtime/src/titan/ast/runtime/AstGuiOutputer.java
package titan.ast.runtime;

import java.util.ArrayList;
import javax.swing.SwingUtilities;

/**
 * 图形化显示树数据结构.
 *
 * @author tian wei jun
 */
class AstGuiOutputer {
  Ast tree;

  AstGuiOutputer(Ast tree) {
    this.tree = tree;
  }

  void output() {
    SwingUtilities.invokeLater(new GuiRunnable(tree));
  }

  private static class GuiRunnable implements Runnable {
    StringTree strTree;

    GuiRunnable(Ast tree) {
      if (null != tree) {
        this.strTree = buildStringTree(tree);
      }
    }

    @Override
    public void run() {
      if (null != strTree) {
        TreeViewerDialog astDialog = new TreeViewerDialog(strTree);
        astDialog.setTitle("ast");
        astDialog.show();
      }
    }

    private StringTree buildStringTree(Ast astTree) {
      StringTree strTree =
          new StringTree(astTree.toString(), new ArrayList<StringTree>(astTree.children.size()));
      if (!astTree.children.isEmpty()) {
        for (Ast astTreeChild : astTree.children) {
          StringTree strTreeChild = buildStringTree(astTreeChild);
          strTree.children.add(strTreeChild);
        }
      }
      return strTree;
    }
  }
}
